### DISCUSSION TITLE
How many times would a (sub)function be evaluated?

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3341] [Q&A] [answered]
Hi there

To explain my equation, let's take this simple constructed example. 
Suppose I have the (sub)function $f(x)$, where $x$ is variable. I then also have the functions 
$g_1(x,p) = pf(x+1)$, 
$g_2(x,p) = pf(x-1)$, 
$g_3(x,p) = f(x+1)/p$, 
$g_4(x,p) = f(x-1)/p$, 
where $p$ is some parameter. I would code this as follows:

```
x = MX.sym('x');
p = MX.sym('p');

% create subfunction f and compute f(x+1) and f(x-1) symbolically 
f = (some expression involving x)
f_func = Function('f',{x},{f});
f_map_func =  f_func.map(2);
f_vals = f_map_func([x+1,x-1]);

% create function g  and compute g_1, g_2, g_3, g_4 symbolically 
a = MX.sym('a');
f_temp = MX.sym('f_temp');
g = a*f_temp ;
g_func = Function('g',{a,f_temp},{g});
g_map_func =  g_func.map(4);
g_vals = g_map_func ([p p 1/p 1/p], [f_vals  f_vals]);

% compute numerical values of g_1, g_2, g_3, g_4
g_vals_func = Function('g_vals',{x,p},{g_vals})
values = g_vals_func(4,10);
```

In this case, how many times would CasADi evaluate $f$? Two times or four times? My intuition on how the expression graph might be built says it will be two times----but I just want to make sure. Thanks!

**Comment by jgillis** [comment, marked as answer]
Two should be feasible. You may check by passing 'print_in' true option when constructing f_func

**Reply by tinusd** [reply]
Thanks a lot for your quick reply! Indeed, it is only called twice:
_Function f (0x21b65995bb0)
Input 0 (i0): 5
Function f (0x21b65995bb0)
Input 0 (i0): 3_


Suppose I change the following:
```
f_map_func =  f_func.map(3);
f_vals = f_map_func([x+1,x-1,x+10]);
g_vals = g_map_func ([p p 1/p 1/p], [f_vals(1:2)  f_vals(1:2)]);
```
So, I symbolically computed three values of $f$, but I only used two of them. From what I see, it will still evaluate $f$ three times:
_Function f (0x21b65995bb0)
Input 0 (i0): 5
Function f (0x21b65995bb0)
Input 0 (i0): 3
Function f (0x21b65995bb0)
Input 0 (i0): 14_

This is the expected behaviour, right? I guess it needs to call f_map_func, which is defined to evaluate $f$ three times---no matter if I only used two values of $f$.

**Reply by jgillis** [reply]
This is expected. You can avoid spurious computations by expanding the overall Function to SX (if your applications allows it). You *may* avoid spurious computations by sticking with MX, but using `always_inline true` for `f`. You can look at generated c code to understand what computations are effectively taking place
